man/mstresourcedump.1
.TH MSTRESOURCEDUMP "1" "March 2020" "mstflint" "User Commands"
.SH NAME
mstresourcedump \- dump resource information
.SH DESCRIPTION
usage: mstresourcedump [\-h] [\-v] {dump,query} ...
.SS "optional arguments:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
Shows tool version
.SS "commands:"
.IP
{dump,query}
.PP
Use 'mstresourcedump <command> \fB\-h\fR' to read about a specific command.
.SH "SEE ALSO"
.B mstflint(1)


debian/changelog
mstflint (4.22.0-1) unstable; urgency=low

  * Updated from MFT-4.22.0

 -- Alex Blago <alexbl@nvidia.com>  Thu, 8 Sep 2022 00:00:00 +0000

mstflint (4.21.0-1) unstable; urgency=low

  * Updated from MFT-4.21.0

 -- Tomer Tubi <ttubi@nvidia.com>  Sun, 31 Jul 2022 00:00:00 +0000

mstflint (4.20.1-1) unstable; urgency=low

  * Updated from MFT-4.20.1

 -- Tomer Tubi <ttubi@nvidia.com>  Tue, 31 May 2022 00:00:00 +0000

mstflint (4.20.0-1) unstable; urgency=low

  * Updated from MFT-4.20.0

 -- Tomer Tubi <ttubi@nvidia.com>  Thu, 28 Apr 2022 00:00:00 +0000

mstflint (4.19.0-1) unstable; urgency=low

  * Updated from MFT-4.19.0 (internal version)

 -- Tomer Tubi <ttubi@nvidia.com>  Thu, 24 Mar 2022 00:00:00 +0000

mstflint (4.18.0-1) unstable; urgency=low

  * Python 2.x - Python 2.x is now end-of-life and no longer supported by MFT.

  * To use the latest and up-to-date MFT tools, we recommend you use Python 3.x.

  * mstflint - When downgrading to a firmware version that does not support the flash type of the device, mstflint will present the user a clear error of such scenario.

  * mstfwreset	Added a new reset-type ("NIC only reset") to mstfwreset which is applicable only to SmartNIC devices.

  * The new reset-type is also the new default for SmartNIC devices. In case of reset-type is set to "NIC only reset", mstfwreset will not reset the internal host.

  * mstlink	- Added support for new PRBS TX/RX patterns (--tx_prbs <tx_prbs_mode> & --rx_prbs <rx_prbs_mode>).

  * For further information, see mstlink Utility section.

  * mstlink - Added new show counters for 16nm devices. To see them run the "show_counter" command.

  * mstlink - Extended the list of the cable information received for 16nm devices when running the "show_module" command.

  * mstlink - Extended the information collection for 7nm and 16nm devices. See "--amber_collect" flag.

  * For further information, see mstlink Utility section.

  * mstlink - Extended the list of the cable information (LOL, LOS, FSM, and module status) for CMIS when running the "show_module" command.

 -- Tomer Tubi <ttubi@nvidia.com>  Thu, 25 Nov 2021 00:00:00 +0200

mstflint (4.17.0-1) unstable; urgency=low

  * Updated from MFT-4.17.0

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Thu, 10 Jun 2021 11:11:11 +0300

mstflint (4.16.0-2) unstable; urgency=low

  * Updated mlxconfig, fixed several bugs

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Mon, 22 Mar 2021 16:58:33 +0300

mstflint (4.16.0-1) unstable; urgency=low

  * Updated from MFT-4.16.0

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Thu, 17 Dec 2020 04:58:33 +0300

mstflint (4.15.0-1) unstable; urgency=low

  * Ported ConnectX6Dx EFUSE programming from MFT 4.16.0

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Thu, 27 Aug 2020 16:50:45 +0300

mstflint (4.15.0-1) unstable; urgency=low

  * Updated from MFT-4.15.0

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Mon, 24 Aug 2020 13:58:33 +0300

mstflint (4.14.0-3) unstable; urgency=low

  * Fixed format-security warning in flint's checkFwVersion.

 -- Dan Goldberg <dang@unknown>  Thu, 18 Jun 2020 00:00:00 +0200

mstflint (4.14.0-2) unstable; urgency=low

  * Updated from MFT-4.14.0. Added BlueField-2 adapter cards support.

 -- Kostiantyn Maksatov <kostiantynm@unknown>  Thu, 11 Jun 2020 15:55:46 +0200

mstflint (4.14.0-1) unstable; urgency=low

  * Updated from MFT-4.14.0

 -- Matan Eliyahu <matanel@unknown>  Tue, 31 Mar 2020 00:00:00 +0200

mstflint (4.13.3-2) unstable; urgency=low

  * Updated from MFT-4.13.3

 -- Matan Eliyahu <matanel@unknown>  Wed, 08 Jan 2020 00:00:00 +0200

mstflint (4.13.3-1) unstable; urgency=low

  * Updated from MFT-4.13.3. Added new tools: mstresourcedump.

 -- Matan Eliyahu <matanel@unknown>  Tue, 31 Dec 2019 00:00:00 +0200

mstflint (4.13.1-1) unstable; urgency=low

  * Updated from MFT-4.13.0

 -- Matan Eliyahu <matanel@unknown>  Thu, 10 Oct 2019 00:00:00 +0200

mstflint (4.13.0-1) unstable; urgency=low

  * Updated from MFT-4.13.0

 -- Matan Eliyahu <matanel@unknown>  Wed, 02 Oct 2019 00:00:00 +0200

mstflint (4.12.0-1) unstable; urgency=low

  * Updated from MFT-4.12.0

 -- Matan Eliyahu <matanel@unknown>  Thu, 13 Jun 2019 00:00:00 +0200

mstflint (4.11.0-5) unstable; urgency=low

  * Updated from MFT-4.11.1

 -- Dan Goldberg <dang@unknown>  Wed, 01 May 2019 00:00:00 +0200

mstflint (4.11.0-4) unstable; urgency=low

  * Updated from MFT-4.11.1

 -- Dan Goldberg <dang@unknown>  Sun, 07 Apr 2019 00:00:00 +0200

mstflint (4.11.0-3) unstable; urgency=low

  * Updated from MFT-4.11.1

 -- Dan Goldberg <dang@unknown>  Mon, 11 Feb 2019 00:00:00 +0200

mstflint (4.11.0-2) unstable; urgency=low

  * Updated from MFT-4.11.0

 -- Dan Goldberg <dang@unknown>  Thu, 31 Jan 2019 00:00:00 +0200

mstflint (4.11.0-1) unstable; urgency=low

  * Updated from MFT-4.11.0. Added new tools: mlxarchive and mstprivhost.

 -- Dan Goldberg <dang@unknown>  Wed, 21 Nov 2018 16:00:00 +0200

mstflint (4.10.0-1) unstable; urgency=low

  * Updated from MFT-4.10.0

 -- Dan Goldberg <dang@unknown>  Sun, 01 Jul 2018 21:02:03 +0200

mstflint (4.10.0-1) unstable; urgency=low

  * Updated from MFT-4.10.0

 -- Adham Masarwah <adham@unknown>  Mon, 02 Apr 2018 15:35:46 +0200

mstflint (4.9.0-1) unstable; urgency=low

  * Updated from MFT-4.9.0

 -- Adham Masarwah <adham@unknown>  Mon, 25 Dec 2017 11:21:46 +0200


mstflint (4.8.0-1) unstable; urgency=low

  * Updated from MFT-4.8.0

 -- Adham Masarwah <adham@unknown>  Wed, 18 Oct 2017 17:52:33 +0200


mstflint (4.7.0-1) unstable; urgency=low

  * Updated from MFT-4.7.0

 -- Adrian Chiris <adrianc@unknown>  Mon, 08 May 2017 11:00:33 +0200


mstflint (4.6.0-1) unstable; urgency=low

  * Updated from MFT-4.6.0

 -- Adham Masarwah <adham@unknown>  Wed, 11 Jan 2017 20:01:33 +0200


mstflint (4.5.0-1) unstable; urgency=low

  * Updated from MFT-4.5.0
  * Notice: as of mstflint-4.6.0 the scrip, hca_ofed_self_test.ofed will no longer be a part of mstflint,

resourcedump/mstresourcedump.py
# Copyright (C) Jan 2020 Mellanox Technologies Ltd. All rights reserved.
# Copyright (c) 2021 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# This software is available to you under a choice of one of two
# licenses.  You may choose to be licensed under the terms of the GNU
# General Public License (GPL) Version 2, available from the file
# COPYING in the main directory of this source tree, or the
# OpenIB.org BSD license below:
#
#     Redistribution and use in source and binary forms, with or
#     without modification, are permitted provided that the following
#     conditions are met:
#
#      - Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      - Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials
#        provided with the distribution.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# --


#######################################################
#
# MlxResDump.py
# Python implementation of the Class MlxResDump
# Generated by Enterprise Architect
# Created on:      14-Aug-2019 10:11:58 AM
# Original author: talve
#
#######################################################
import sys
import os

if sys.version_info[0] < 3:
    print("Error: This tool supports python 3.x only. Exiting...")
    exit(1)

import argparse
from commands.CommandFactory import CommandFactory
from utils import constants as cs
sys.path.append(os.path.join("common"))
import tools_version


class MlxResDump:
    """This class is responsible for the resource dump UI by handling the user inputs and
       and running the right command.
    """
    @staticmethod
    def _decimal_hex_check(inp):
        """This method check if the string input is hex or decimal.
           if the input is hex the method convert it to decimal number
           otherwise convert it to decimal number.
        """
        try:
            if inp.startswith("0x"):
                inp = int(inp, 16)
            else:
                inp = int(inp)
            return inp
        except Exception as _:
            msg = "wrong value: {0} is not decimal or hex".format(inp)
            raise argparse.ArgumentTypeError(msg)

    @staticmethod
    def _decimal_hex_to_str_hex(inp):
        """This method check if the string input is hex or decimal.
           and convert it to hex number.
           in case that the input is not a number, the method will
           return the input as is (str).
        """
        try:
            if inp.startswith("0x"):
                inp = hex(int(inp, 16))
            else:
                inp = hex(int(inp))
            return str(inp)
        except Exception as _:
            return inp

    @staticmethod
    def _num_of_objs_check(inp):
        """This method check if the num of objects parameter is valid
        """
        if inp in (['all', 'active']):
            if inp == 'all':
                inp = cs.NUM_OF_OBJ_ALL
            else:
                inp = cs.NUM_OF_OBJ_ACTIVE
            return inp
        else:
            try:
                if inp.startswith("0x"):
                    inp = int(inp, 16)
                else:
                    inp = int(inp)
                return inp
            except Exception as _:
                msg = "numOfObj accepts the following values: ['all', 'active', number]"
                raise argparse.ArgumentTypeError(msg)

    @staticmethod
    def _depth_check(inp):
        """This method check if the num depth parameter is valid
        """
        if inp == "inf":
            return inp
        else:
            try:
                inp = int(inp)
                return inp
            except Exception as _:
                msg = "depth accepts the following values: ['inf', number]"
                raise argparse.ArgumentTypeError(msg)

    def run(self):
        # main parser
        tool_name = os.path.basename(__file__.split('.')[0])
        parser = argparse.ArgumentParser(prog=tool_name,
                                         epilog="Use '{0} <command> -h' to read about a specific command.".format(tool_name),
                                         formatter_class=argparse.RawTextHelpFormatter)
        parser.add_argument('-v', '--version', action='version', help='Shows tool version',
                            version=tools_version.GetVersionString(tool_name, None))

        # commands sub parser
        commands = parser.add_subparsers(title='commands')

        # dump sub parser
        dump_parser = commands.add_parser(cs.RESOURCE_DUMP_COMMAND_TYPE_DUMP)
        dump_parser.set_defaults(parser=cs.RESOURCE_DUMP_COMMAND_TYPE_DUMP)

        # required arguments by dump sub parser
        dump_required_args = dump_parser.add_argument_group('required arguments')
        dump_required_args.add_argument(cs.UI_DASHES + cs.UI_ARG_DEVICE, cs.UI_DASHES_SHORT + cs.UI_ARG_DEVICE_SHORT,
                                        help='The device name', required=True)
        dump_required_args.add_argument(cs.UI_DASHES + cs.UI_ARG_SEGMENT, help='The segment to dump', required=True,
                                        type=self._decimal_hex_to_str_hex)

        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_VHCAID.replace("_", "-"),
                                 help='The virtual HCA (host channel adapter, NIC) ID')
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_INDEX1,
                                 help='The first context index to dump (if supported for this segment)',
                                 type=self._decimal_hex_check)
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_INDEX2,
                                 help='The second context index to dump (if supported for this segment)',
                                 type=self._decimal_hex_check)
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_NUMOFOBJ1.replace("_", "-"),
                                 help='The number of objects to be dumped (if supported for this segment). accepts: ["all", "active", number, depends on the capabilities]',
                                 type=self._num_of_objs_check)
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_NUMOFOBJ2.replace("_", "-"),
                                 help='The number of objects to be dumped (if supported for this segment). accepts: ["all", "active", number, depends on the capabilities]',
                                 type=self._num_of_objs_check)
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_DEPTH,
                                 help='The depth of walking through reference segments. 0 stands for flat, '
                                      '1 allows crawling of a single layer down the struct, etc. "inf" for all',
                                 type=self._depth_check)
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_BIN,
                                 help='The output to a binary file that replaces the default print in hexadecimal,'
                                      ' a readable format')
        dump_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_MEM.replace("_", "-"), default=None,
                                 help='Perform the dump through memory (ofed with rdma-core dependency). accepts: [rdma device (for example "mlx5_4")]')

        # query sub parser
        query_parser = commands.add_parser(cs.RESOURCE_DUMP_COMMAND_TYPE_QUERY)
        query_parser.set_defaults(parser=cs.RESOURCE_DUMP_COMMAND_TYPE_QUERY)
        query_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_VHCAID.replace("_", "-"),
                                  help='The virtual HCA (host channel adapter, NIC) ID')
        query_parser.add_argument(cs.UI_DASHES + cs.UI_ARG_MEM.replace("_", "-"), default=None,
                                  help='Perform the dump through memory (ofed with rdma-core dependency). accepts: [rdma device (for example "mlx5_4")]')

        # required arguments by query sub parser
        query_required_args = query_parser.add_argument_group('required arguments')
        query_required_args.add_argument(cs.UI_DASHES + cs.UI_ARG_DEVICE, cs.UI_DASHES_SHORT + cs.UI_ARG_DEVICE_SHORT,
                                         help='The device name', required=True)

        # args = parser.parse_args(
        #     'dump -d test --source test2 --vHCAid 15 --index 1 --index2 2 --numOfObj1 10 --numOfObj2 20
        #           --depth inf --bin tt.txt'.split())
        arguments = parser.parse_args()

        if len(sys.argv) < 2:
            parser.print_usage()
            sys.exit(1)

        return arguments


def create_command(arguments):
    """This method creates the right command.
    """
    try:
        command_type = arguments.parser
        created_command = CommandFactory.create(command_type, **vars(arguments))
        return created_command
    except Exception as exp:
        raise Exception("failed to create command of type: {0} with exception: {1}".format(command_type, exp))


if __name__ == '__main__':
    try:
        args = MlxResDump().run()
        command = create_command(args)
        command.execute()
    except Exception as e:
        print("Error: {0}. Exiting...".format(e))
        sys.exit(1)
